url.c

```diff
@@ -1567,12 +1570,104 @@ int is_gpg_signed(char *file, int check)
     fprintf(stderr, "%s: gpg signature %s\n", file, err ? "failed" : "ok");
   }
 
+  if(config.debug) fprintf(stderr, "%s: gpg check = %d\n", file, err);
+
+  return err;
+}
+
+
+/*
+  Test if 'file' is a signed rpm.
+  If so, verify signature.
+
+  Return values:
+    -1: file or 'rpmkeys' not found
+     0: file rpm format, sig ok
+     1: file rpm format, sig wrong
+     2: file not rpm format or not signed
+*/
+int is_rpm_signed(char *file)
+{
+  char *cmd = NULL, *buf = NULL;
+  int err = -1, is_sig = 0, sig_ok = 0;
+  size_t len = 0;
+  FILE *f;
+
+  if(util_check_exist(file) != 'r') {
+    if(config.debug) fprintf(stderr, "%s: rpm sig check = %d\n", file, err);
+
+    return err;
+  }
+
+  char *type = util_fstype(file, NULL);
+  if(!type || strcmp(type, "rpm")) return 2;
+
+  strprintf(&cmd, "rpmkeys --checksig --define '%%_keyringpath /pubkeys' '%s' 2>&1", file);
+
+  if((f = popen(cmd, "r"))) {
+    while(getline(&buf, &len, f) > 0) {
+      char *s = strrchr(buf, ':') ?: buf;
+
+      if(config.debug >= 2) fprintf(stderr, "%s", buf);
+
+      if(strcasestr(s, " pgp ") || strcasestr(s, " gpg ")) is_sig = 1;
+      if(strstr(s, " pgp ") || strstr(s, " gpg ")) sig_ok = 1;
+    }
+    err = pclose(f) ? 1 : 0;
+    if(config.debug >= 2) fprintf(stderr, "rpmkeys returned %s\n", err ? "an error" : "ok");
+  }
+
+  str_copy(&cmd, NULL);
+  free(buf);
+
+  if(err != -1) {
```

Comment:
when it can happen? for me only if popen failed or getline is 0. In such case buf can be NULL and I think it is not needed be free. so it can be shortly exited above with `if (err == -1) return err;`